File: Sources/SwiftMTH/MSDPManager.swift
// MSDP protocol control bytes (shared with MSDPConverter.swift)
private let MSDP_VAR: UInt8 = 1
private let MSDP_VAL: UInt8 = 2
private let MSDP_TABLE_OPEN: UInt8 = 3
private let MSDP_TABLE_CLOSE: UInt8 = 4
private let MSDP_ARRAY_OPEN: UInt8 = 5
private let MSDP_ARRAY_CLOSE: UInt8 = 6

private let IAC: UInt8 = 255
private let SB: UInt8 = 250
private let SE: UInt8 = 240
private let TELOPT_MSDP: UInt8 = 69

/// Callback for writing MSDP data to the connection.
public typealias MSDPWriteHandler = ([UInt8]) -> Void

/// Callback for logging.
public typealias MSDPLogHandler = (String) -> Void

/// Manages MSDP variable state for a single connection.
///
/// Replaces the C per-connection `msdp_data` array and the global `msdp_table`
/// functions. Each `MSDPManager` instance corresponds to one telnet session.
///
/// Not `Sendable` — intended for use on a single thread/connection.
public final class MSDPManager {
    /// Variable definitions (shared, immutable).
    private let definitions: [MSDPVariableDefinition]

    /// Fast lookup from variable name to index in `definitions`.
    private let nameIndex: [String: Int]

    /// Per-connection variable state, indexed by position in `definitions`.
    private var state: [MSDPVariableState]

    /// Whether any variable needs to be flushed via `flushUpdates()`.
    public private(set) var needsUpdate: Bool = false

    /// Whether this connection uses GMCP (JSON) instead of raw MSDP.
    public var usesGMCP: Bool = false

    /// Called to write bytes to the connection.
    public let writeHandler: MSDPWriteHandler

    /// Called for log messages.
    public let logHandler: MSDPLogHandler

    public init(
        table: [MSDPVariableDefinition] = defaultMSDPTable,
        writeHandler: @escaping MSDPWriteHandler,
        logHandler: @escaping MSDPLogHandler = { _ in }
    ) {
        self.definitions = table
        var index: [String: Int] = [:]
        index.reserveCapacity(table.count)
        for (i, def) in table.enumerated() {
            index[def.name] = i
        }
        self.nameIndex = index
        self.state = table.map { def in
            MSDPVariableState(value: "", flags: def.flags)
        }
        self.writeHandler = writeHandler
        self.logHandler = logHandler
    }

    // MARK: - Variable Lookup

    /// Find the index of a variable by name. Returns nil if not found.
    private func findIndex(_ name: String) -> Int? {
        nameIndex[name]
    }

    // MARK: - Reading

    /// Get the current value of a variable, or nil if unknown.
    public func getVariable(_ name: String) -> String? {
        guard let idx = findIndex(name) else {
            logHandler("msdp_get_var: Unknown variable: \(name).")
            return nil
        }
        return state[idx].value
    }

    // MARK: - Updating

    /// Update a variable's value. If the variable is being reported and the
    /// value changed, queues it for the next `flushUpdates()` call.
    ///
    /// Equivalent to C `msdp_update_var`.
    public func updateVariable(_ name: String, value: String) {
        guard let idx = findIndex(name) else {
            logHandler("msdp_update_var: Unknown variable: \(name).")
            return
        }

        if state[idx].value != value {
            if state[idx].flags.contains(.reported) {
                state[idx].flags.insert(.updated)
                needsUpdate = true
            }
            state[idx].value = value
        }
    }

    /// Update a variable's value and immediately send it if reported.
    ///
    /// Equivalent to C `msdp_update_var_instant`.
    public func updateVariableImmediate(_ name: String, value: String) {
        guard let idx = findIndex(name) else {
            logHandler("msdp_update_var_instant: Unknown variable: \(name).")
            return
        }

        if state[idx].value != value {
            state[idx].value = value
        }

        if state[idx].flags.contains(.reported) {
            var packet: [UInt8] = [IAC, SB, TELOPT_MSDP]
            packet.append(MSDP_VAR)
            packet.append(contentsOf: definitions[idx].name.utf8)
            packet.append(MSDP_VAL)
            packet.append(contentsOf: value.utf8)
            packet.append(IAC)
            packet.append(SE)
            writeMSDP(packet)
        }
    }

    /// Send all reported variables that have been updated since the last flush.
    ///
    /// Equivalent to C `msdp_send_update`.
    public func flushUpdates() {
        var packet: [UInt8] = [IAC, SB, TELOPT_MSDP]
        var hasContent = false

        for idx in 0..<definitions.count {
            if state[idx].flags.contains(.updated) {
                packet.append(MSDP_VAR)
                packet.append(contentsOf: definitions[idx].name.utf8)
                packet.append(MSDP_VAL)
                packet.append(contentsOf: state[idx].value.utf8)
                state[idx].flags.remove(.updated)
                hasContent = true
            }
        }

        packet.append(IAC)
        packet.append(SE)

        if hasContent {
            writeMSDP(packet)
        }

        needsUpdate = false
    }

    // MARK: - Client Commands

    /// Process an incoming MSDP variable/value pair from the client.
    ///
    /// Equivalent to C `process_msdp_varval`.
    public func processVarVal(_ variable: String, value: String) {
        guard let varIdx = findIndex(variable) else { return }
        let def = definitions[varIdx]

        if def.flags.contains(.configurable) {
            state[varIdx].value = value
            handleCommand(varIdx)
            return
        }

        if def.flags.contains(.command) {
            if !value.isEmpty && value.utf8.first == MSDP_ARRAY_OPEN {
                processArray(varIdx, value: value)
            } else {
                guard let valIdx = findIndex(value) else { return }
                handleCommandWithArgument(varIdx, argumentIndex: valIdx)
            }
        }
    }

    /// Process a 1D array argument for a command variable.
    private func processArray(_ varIdx: Int, value: String) {
        let bytes = Array(value.utf8)
        var i = 0
        var buf: [UInt8] = []

        while i < bytes.count {
            switch bytes[i] {
            case MSDP_ARRAY_OPEN:
                i += 1
            case MSDP_VAL:
                if !buf.isEmpty {
                    let name = String(decoding: buf, as: UTF8.self)
                    if let argIdx = findIndex(name) {
                        handleCommandWithArgument(varIdx, argumentIndex: argIdx)
                    }
                }
                buf.removeAll()
                i += 1
            case MSDP_ARRAY_CLOSE:
                if !buf.isEmpty {
                    let name = String(decoding: buf, as: UTF8.self)
                    if let argIdx = findIndex(name) {
                        handleCommandWithArgument(varIdx, argumentIndex: argIdx)
                    }
                }
                return
            default:
                buf.append(bytes[i])
                i += 1
            }
        }

        if !buf.isEmpty {
            let name = String(decoding: buf, as: UTF8.self)
            if let argIdx = findIndex(name) {
                handleCommandWithArgument(varIdx, argumentIndex: argIdx)
            }
        }
    }

    /// Dispatch a command with no argument (configurable variable handler).
    private func handleCommand(_ varIdx: Int) {
        let name = definitions[varIdx].name
        switch name {
        case "ARACHNOS_DEVEL", "ARACHNOS_MUDLIST":
            // Arachnos handlers are MUD-specific; delegate up
            break
        default:
            break
        }
    }

    /// Dispatch a command with a variable-index argument.
    ///
    /// The C code uses function pointers; we use a switch on the command name.
    private func handleCommandWithArgument(_ cmdIdx: Int, argumentIndex argIdx: Int) {
        let name = definitions[cmdIdx].name
        switch name {
        case "LIST":
            commandList(argIdx)
        case "REPORT":
            commandReport(argIdx)
        case "UNREPORT":
            commandUnreport(argIdx)
        case "SEND":
            commandSend(argIdx)
        case "RESET":
            commandReset(argIdx)
        default:
            break
        }
    }

    // MARK: - Command Implementations

    /// LIST command: send a list of variables matching the requested category.
    ///
    /// Equivalent to C `msdp_command_list`.
    private func commandList(_ index: Int) {
        let def = definitions[index]
        guard def.flags.contains(.list) else { return }

        var packet: [UInt8] = [IAC, SB, TELOPT_MSDP]
        packet.append(MSDP_VAR)
        packet.append(contentsOf: def.name.utf8)
        packet.append(MSDP_VAL)
        packet.append(MSDP_ARRAY_OPEN)

        let flag = def.flags.subtracting(.list)

        for idx in 0..<definitions.count {
            if !flag.isEmpty {
                // List variables matching the flag (excluding other list vars)
                if state[idx].flags.contains(flag) && !state[idx].flags.contains(.list) {
                    packet.append(MSDP_VAL)
                    packet.append(contentsOf: definitions[idx].name.utf8)
                }
            } else {
                // flag is empty after removing .list -> this is "LISTS", list all list variables
                if state[idx].flags.contains(.list) {
                    packet.append(MSDP_VAL)
                    packet.append(contentsOf: definitions[idx].name.utf8)
                }
            }
        }

        packet.append(MSDP_ARRAY_CLOSE)
        packet.append(IAC)
        packet.append(SE)

        writeMSDP(packet)
    }

    /// REPORT command: enable auto-reporting for a variable.
    ///
    /// Equivalent to C `msdp_command_report`.
    private func commandReport(_ index: Int) {
        guard definitions[index].flags.contains(.reportable) else { return }

        state[index].flags.insert(.reported)

        guard definitions[index].flags.contains(.sendable) else { return }

        state[index].flags.insert(.updated)
        needsUpdate = true
    }

    /// UNREPORT command: disable auto-reporting for a variable.
    ///
    /// Equivalent to C `msdp_command_unreport`.
    private func commandUnreport(_ index: Int) {
        guard definitions[index].flags.contains(.reportable) else { return }
        state[index].flags.remove(.reported)
    }

    /// SEND command: queue a sendable variable for immediate update.
    ///
    /// Equivalent to C `msdp_command_send`.
    private func commandSend(_ index: Int) {
        if state[index].flags.contains(.sendable) {
            state[index].flags.insert(.updated)
            needsUpdate = true
        }
    }

    /// RESET command: reset all variables matching the list's flag.
    ///
    /// Equivalent to C `msdp_command_reset`.
    private func commandReset(_ index: Int) {
        guard definitions[index].flags.contains(.list) else { return }

        let flag = definitions[index].flags.subtracting(.list)

        for idx in 0..<definitions.count {
            if state[idx].flags.contains(flag) {
                state[idx].flags = definitions[idx].flags
            }
        }
    }

    // MARK: - Output

    /// Write an MSDP packet, converting to GMCP JSON if needed.
    ///
    /// Equivalent to C `write_msdp_to_descriptor`.
    private func writeMSDP(_ packet: [UInt8]) {
        if usesGMCP {
            let json = msdp2json(packet)
            writeHandler(json)
        } else {
            writeHandler(packet)
        }
    }
}

File: kotlin/mth-core/src/main/kotlin/mth/core/MSDPManager.kt
package mth.core

/**
 * Manages MSDP variable state for a single connection.
 *
 * Replaces the C per-connection `msdp_data` array and the global `msdp_table`
 * functions. Each MSDPManager instance corresponds to one telnet session.
 */
class MSDPManager(
    table: List<MSDPVariableDefinition> = defaultMSDPTable,
    val writeHandler: (ByteArray) -> Unit,
    val logHandler: (String) -> Unit = {}
) {
    // MSDP protocol control bytes
    companion object {
        private const val MSDP_VAR: Byte = 1
        private const val MSDP_VAL: Byte = 2
        private const val MSDP_TABLE_OPEN: Byte = 3
        private const val MSDP_TABLE_CLOSE: Byte = 4
        private const val MSDP_ARRAY_OPEN: Byte = 5
        private const val MSDP_ARRAY_CLOSE: Byte = 6
        private val IAC: Byte = 0xFF.toByte()
        private val SB: Byte = 0xFA.toByte()
        private val SE: Byte = 0xF0.toByte()
        private const val TELOPT_MSDP: Byte = 69
    }

    /** Variable definitions (shared, immutable). */
    private val definitions: List<MSDPVariableDefinition> = table

    /** Fast lookup from variable name to index in definitions. */
    private val nameIndex: Map<String, Int> = buildMap {
        for ((i, def) in table.withIndex()) {
            put(def.name, i)
        }
    }

    /** Per-connection variable state, indexed by position in definitions. */
    private val state: MutableList<MSDPVariableState> = table.map { def ->
        MSDPVariableState(value = "", flags = def.flags)
    }.toMutableList()

    /** Whether any variable needs to be flushed via flushUpdates(). */
    var needsUpdate: Boolean = false
        private set

    /** Whether this connection uses GMCP (JSON) instead of raw MSDP. */
    var usesGMCP: Boolean = false

    // -- Variable Lookup --

    private fun findIndex(name: String): Int? = nameIndex[name]

    // -- Copyover Support --

    /** Names of variables currently marked as reported (for copyover snapshot). */
    val reportedVariableNames: List<String>
        get() {
            val names = mutableListOf<String>()
            for (idx in definitions.indices) {
                if (MSDPFlags.REPORTED in state[idx].flags) {
                    names.add(definitions[idx].name)
                }
            }
            return names
        }

    /** Restore reported variable flags from a list of variable names (after copyover). */
    fun restoreReportedVariables(names: List<String>) {
        for (name in names) {
            val idx = findIndex(name) ?: continue
            state[idx].flags = state[idx].flags.insert(MSDPFlags.REPORTED)
        }
    }

    // -- Reading --

    /** Get the current value of a variable, or null if unknown. */
    fun getVariable(name: String): String? {
        val idx = findIndex(name)
        if (idx == null) {
            logHandler("msdp_get_var: Unknown variable: $name.")
            return null
        }
        return state[idx].value
    }

    // -- Updating --

    /**
     * Update a variable's value. If the variable is being reported and the
     * value changed, queues it for the next flushUpdates() call.
     */
    fun updateVariable(name: String, value: String) {
        val idx = findIndex(name)
        if (idx == null) {
            logHandler("msdp_update_var: Unknown variable: $name.")
            return
        }

        if (state[idx].value != value) {
            if (MSDPFlags.REPORTED in state[idx].flags) {
                state[idx].flags = state[idx].flags.insert(MSDPFlags.UPDATED)
                needsUpdate = true
            }
            state[idx].value = value
        }
    }

    /**
     * Update a variable's value and immediately send it if reported.
     */
    fun updateVariableImmediate(name: String, value: String) {
        val idx = findIndex(name)
        if (idx == null) {
            logHandler("msdp_update_var_instant: Unknown variable: $name.")
            return
        }

        if (state[idx].value != value) {
            state[idx].value = value
        }

        if (MSDPFlags.REPORTED in state[idx].flags) {
            val packet = mutableListOf<Byte>()
            packet.add(IAC)
            packet.add(SB)
            packet.add(TELOPT_MSDP)
            packet.add(MSDP_VAR)
            packet.addAll(definitions[idx].name.toByteArray(Charsets.UTF_8).toList())
            packet.add(MSDP_VAL)
            packet.addAll(value.toByteArray(Charsets.UTF_8).toList())
            packet.add(IAC)
            packet.add(SE)
            writeMSDP(packet.toByteArray())
        }
    }

    /**
     * Send all reported variables that have been updated since the last flush.
     */
    fun flushUpdates() {
        val packet = mutableListOf<Byte>()
        packet.add(IAC)
        packet.add(SB)
        packet.add(TELOPT_MSDP)
        var hasContent = false

        for (idx in definitions.indices) {
            if (MSDPFlags.UPDATED in state[idx].flags) {
                packet.add(MSDP_VAR)
                packet.addAll(definitions[idx].name.toByteArray(Charsets.UTF_8).toList())
                packet.add(MSDP_VAL)
                packet.addAll(state[idx].value.toByteArray(Charsets.UTF_8).toList())
                state[idx].flags = state[idx].flags.remove(MSDPFlags.UPDATED)
                hasContent = true
            }
        }

        packet.add(IAC)
        packet.add(SE)

        if (hasContent) {
            writeMSDP(packet.toByteArray())
        }

        needsUpdate = false
    }

    // -- Client Commands --

    /**
     * Process an incoming MSDP variable/value pair from the client.
     */
    fun processVarVal(variable: String, value: String) {
        val varIdx = findIndex(variable) ?: return
        val def = definitions[varIdx]

        if (MSDPFlags.CONFIGURABLE in def.flags) {
            state[varIdx].value = value
            handleCommand(varIdx)
            return
        }

        if (MSDPFlags.COMMAND in def.flags) {
            if (value.isNotEmpty() && value.toByteArray(Charsets.UTF_8)[0] == MSDP_ARRAY_OPEN) {
                processArray(varIdx, value)
            } else {
                val valIdx = findIndex(value) ?: return
                handleCommandWithArgument(varIdx, valIdx)
            }
        }
    }

    /** Process a 1D array argument for a command variable. */
    private fun processArray(varIdx: Int, value: String) {
        val bytes = value.toByteArray(Charsets.UTF_8)
        var i = 0
        val buf = mutableListOf<Byte>()

        while (i < bytes.size) {
            when (bytes[i]) {
                MSDP_ARRAY_OPEN -> {
                    i++
                }
                MSDP_VAL -> {
                    if (buf.isNotEmpty()) {
                        val name = String(buf.toByteArray(), Charsets.UTF_8)
                        val argIdx = findIndex(name)
                        if (argIdx != null) {
                            handleCommandWithArgument(varIdx, argIdx)
                        }
                    }
                    buf.clear()
                    i++
                }
                MSDP_ARRAY_CLOSE -> {
                    if (buf.isNotEmpty()) {
                        val name = String(buf.toByteArray(), Charsets.UTF_8)
                        val argIdx = findIndex(name)
                        if (argIdx != null) {
                            handleCommandWithArgument(varIdx, argIdx)
                        }
                    }
                    return
                }
                else -> {
                    buf.add(bytes[i])
                    i++
                }
            }
        }

        if (buf.isNotEmpty()) {
            val name = String(buf.toByteArray(), Charsets.UTF_8)
            val argIdx = findIndex(name)
            if (argIdx != null) {
                handleCommandWithArgument(varIdx, argIdx)
            }
        }
    }

    /** Dispatch a command with no argument (configurable variable handler). */
    private fun handleCommand(varIdx: Int) {
        @Suppress("UNUSED_VARIABLE")
        val name = definitions[varIdx].name
        // Arachnos handlers are MUD-specific; delegate up
    }

    /** Dispatch a command with a variable-index argument. */
    private fun handleCommandWithArgument(cmdIdx: Int, argumentIndex: Int) {
        when (definitions[cmdIdx].name) {
            "LIST" -> commandList(argumentIndex)
            "REPORT" -> commandReport(argumentIndex)
            "UNREPORT" -> commandUnreport(argumentIndex)
            "SEND" -> commandSend(argumentIndex)
            "RESET" -> commandReset(argumentIndex)
        }
    }

    // -- Command Implementations --

    /** LIST command: send a list of variables matching the requested category. */
    private fun commandList(index: Int) {
        val def = definitions[index]
        if (MSDPFlags.LIST !in def.flags) return

        val packet = mutableListOf<Byte>()
        packet.add(IAC)
        packet.add(SB)
        packet.add(TELOPT_MSDP)
        packet.add(MSDP_VAR)
        packet.addAll(def.name.toByteArray(Charsets.UTF_8).toList())
        packet.add(MSDP_VAL)
        packet.add(MSDP_ARRAY_OPEN)

        val flag = def.flags.subtract(MSDPFlags.LIST)

        for (idx in definitions.indices) {
            if (!flag.isEmpty()) {
                // List variables matching the flag (excluding other list vars)
                if (flag in state[idx].flags && MSDPFlags.LIST !in state[idx].flags) {
                    packet.add(MSDP_VAL)
                    packet.addAll(definitions[idx].name.toByteArray(Charsets.UTF_8).toList())
                }
            } else {
                // flag is empty after removing .list -> this is "LISTS", list all list variables
                if (MSDPFlags.LIST in state[idx].flags) {
                    packet.add(MSDP_VAL)
                    packet.addAll(definitions[idx].name.toByteArray(Charsets.UTF_8).toList())
                }
            }
        }

        packet.add(MSDP_ARRAY_CLOSE)
        packet.add(IAC)
        packet.add(SE)

        writeMSDP(packet.toByteArray())
    }

    /** REPORT command: enable auto-reporting for a variable. */
    private fun commandReport(index: Int) {
        if (MSDPFlags.REPORTABLE !in definitions[index].flags) return

        state[index].flags = state[index].flags.insert(MSDPFlags.REPORTED)

        if (MSDPFlags.SENDABLE !in definitions[index].flags) return

        state[index].flags = state[index].flags.insert(MSDPFlags.UPDATED)
        needsUpdate = true
    }

    /** UNREPORT command: disable auto-reporting for a variable. */
    private fun commandUnreport(index: Int) {
        if (MSDPFlags.REPORTABLE !in definitions[index].flags) return
        state[index].flags = state[index].flags.remove(MSDPFlags.REPORTED)
    }

    /** SEND command: queue a sendable variable for immediate update. */
    private fun commandSend(index: Int) {
        if (MSDPFlags.SENDABLE in state[index].flags) {
            state[index].flags = state[index].flags.insert(MSDPFlags.UPDATED)
            needsUpdate = true
        }
    }

    /** RESET command: reset all variables matching the list's flag. */
    private fun commandReset(index: Int) {
        if (MSDPFlags.LIST !in definitions[index].flags) return

        val flag = definitions[index].flags.subtract(MSDPFlags.LIST)

        for (idx in definitions.indices) {
            if (flag in state[idx].flags) {
                state[idx].flags = definitions[idx].flags
            }
        }
    }

    // -- Output --

    /** Write an MSDP packet, converting to GMCP JSON if needed. */
    private fun writeMSDP(packet: ByteArray) {
        if (usesGMCP) {
            val json = msdp2json(packet)
            writeHandler(json)
        } else {
            writeHandler(packet)
        }
    }
}
